### DISCUSSION TITLE
usePrepareContractWrite fails for payable contract if args has many key-value pairs

### DISCUSSION THREAD

**Discussion by Lactantius** [wevm/wagmi#1611] [Q&A]
I have a contract function that takes several values in a struct, and I want to interact with it with `usePrepareContractWrite`:

```typescript
const { config } = usePrepareContractWrite({
  address: contractAddress,
  abi: [
    {
      inputs: [
        {
          components: [
            {
              internalType: "uint256",
              name: "tokenAmount",
              type: "uint256",
            },
            {
              internalType: "string",
              name: "tokenCode",
              type: "string",
            },
            // More arguments...
          ],
          internalType: "struct Test",
          name: "test",
          type: "tuple",
        },
      ],
      name: "test",
      outputs: [
        {
          internalType: "uint256",
          name: "",
          type: "uint256",
        },
      ],
      stateMutability: "payable",
      type: "function",
    },
  ],
  functionName: "test",
  args: [
    {
      tokenAmount: Utils.parseEther(String(formik.values.amount)),
      tokenCode: "MATIC",
      // More arguments...
    },
  ],
  overrides: {
    value: Utils.parseEther(String(formik.values.amount)),
    gasLimit: 500000,
  },
});
console.log(formik.values.amount);
console.log(config);
const { write } = useContractWrite(config);
console.log(write);
```

This will work as written, but when I start adding more key-value pairs to `args` then I get `invalid BigNumber value (argument="value", value=undefined, code=INVALID_ARGUMENT, version=bignumber/5.7.0`, and `write` is `undefined` unless `value` and `tokenAmount` are `0`. I can always use two values in `args`, but the exact number I can add before failure seems to vary.

This is using Next and Wagmi 0.9.6 (using older version because of ConnectKit) and Next 13.0.6.

**Comment by tmm** [comment]
Everything looks correct. I would double check that `Utils.parseEther(String(formik.values.amount))` is doing what you think it is. The error you posted makes it seems like an error is happening at that point with ethers.

**Reply by Lactantius** [reply]
Thanks. Logging the request when `amount` is `1` gives `value: BigNumber {_hex: '0x0de0b6b3a7640000', _isBigNumber: true}`.
Logging `Utils.parseEther(String(formik.values.amount)).toString();` gives `1000000000000000000`.

**Reply by apoorvlathey** [reply]
I stumbled upon the same error. The actual cause for it was my args' struct was defined incorrectly, but the error referred to txn's  `value`  being wrong for some reason.
It would have been faster for me to debug if the error was accurate.

**Reply by tmm** [reply]
Agree about error messages. Unfortunately this is an ethers error. Have a major update coming soon that will make error messages (and a load of other things) significantly better.

**Reply by Lactantius** [reply]
Thanks. I did get it working properly with plain ethers; I will double-check what I had for Wagmi.
